### PR TITLE
Add dallas multisensor support

### PR DESCRIPTION
**Pull request by lobo0**
Add possibility to use up to 8 dallas sensors at one pin in usermode - DALLASTEMPERATURE.
Sensors are automaticaly discovered. 
Also there is a possibility to send the information by mqtt and show information about temperature on info section .

### CODE REVIEW

**Comment by blazoncek** [comment]
Thank you for this contribution. I was postponing fixing this usermod for too long and it looks like it paid off. 😀 
One question: Do you see occasional incorrect reading on ESP32 or does your code solve that issue?

I will pull this into my fork and test it then merge if everything is ok.

**Comment by lobo0** [comment]
I don't test it on esp32, but on esp8266 I can see occasional incorrect reading (85*C). 
I'll try not to send erroneous readings.

**Comment by blazoncek** [comment]
Please do test on ESP32 as well since there are issues on it.

**Comment by lobo0** [comment]
Currently I don't have any ESP32, but I think it should work. Could anyone test this code on ESP32?

**Comment by Aircoookie** [comment]
Does not appear to work for me on my in-field ESP32, `DS18X20Data.sensors` is 0.
Will test with another unit tomorrow, as my main unit may have a hardware issue with the temp sensor (it stopped working some time ago)

**Comment by blazoncek** [comment]
Does not work for me as well on ESP32.
Using @srg74 Wemos shield (confirmed working prior to this change) with DS18B20 at GPIO23.

File: usermods/Temperature/usermod_temperature.h [hunk 1]
```diff
@@ -12,17 +12,47 @@
...
+struct {
+  float temp;
+  uint16_t numread;
+  uint8_t address[8];
+  uint8_t index;
+  uint8_t valid;
+} ds18x20_sensor[DS18X20_MAX_SENSORS];
+
+struct {
+  char name[17];
```
Comment (blazoncek) [inline usermods/Temperature/usermod_temperature.h:45]: Why name? Waste of RAM.

File: usermods/Temperature/usermod_temperature.h [hunk 2]
```diff
@@ -12,17 +12,47 @@
...
+
+uint8_t ds18x20_chipids[] = { 0, DS18S20_CHIPID, DS1822_CHIPID, DS18B20_CHIPID, MAX31850_CHIPID };
+
+struct {
+  float temp;
+  uint16_t numread;
+  uint8_t address[8];
+  uint8_t index;
+  uint8_t valid;
+} ds18x20_sensor[DS18X20_MAX_SENSORS];
```
Comment (blazoncek) [inline usermods/Temperature/usermod_temperature.h:42]: Why not dynamically allocate from heap as much as there are sensors?

File: usermods/Temperature/usermod_temperature.h [hunk 3]
```diff
@@ -52,80 +82,110 @@ class UsermodTemperature : public Usermod {
...
+            (ds18x20_sensor[num_sensors].address[0] == DS1822_CHIPID) ||
+            (ds18x20_sensor[num_sensors].address[0] == DS18B20_CHIPID) ||
+            (ds18x20_sensor[num_sensors].address[0] == MAX31850_CHIPID))) {
+          num_sensors++;
+        }
+      }
+      for (uint32_t i = 0; i < num_sensors; i++) {
+        ds18x20_sensor[i].index = i;
+      }
+      for (uint32_t i = 0; i < num_sensors; i++) {
```
Comment (blazoncek) [inline usermods/Temperature/usermod_temperature.h:112]: Is this necessary? Why swapping?

File: usermods/Temperature/usermod_temperature.h [hunk 4]
```diff
@@ -52,80 +82,110 @@ class UsermodTemperature : public Usermod {
...
-
-    void readTemperature() {
-      temperature = readDallas();
-      lastMeasurement = millis();
-      waitingForConversion = false;
-      //DEBUG_PRINTF("Read temperature %2.1f.\n", temperature); // does not work properly on 8266
-      DEBUG_PRINT(F("Read temperature "));
-      DEBUG_PRINTLN(temperature);
+    
+    void Ds18x20Search(void) {
```
Comment (blazoncek) [inline usermods/Temperature/usermod_temperature.h:91]: ESP32 may have timing issues with OneWire. Check this routine thoroughly on ESP32 with short and long (3-5m) wires on sensors.

**Review by blazoncek — CHANGES_REQUESTED** [review]
Does not work on ESP32 since it cannot find a sensor.

**Comment by blazoncek** [comment]
Any updates? I was not able to make it work on ESP32.

**Comment by lobo0** [comment]
I am waiting for ESP32 (it's on the way), so I need more time to test it.

**Comment by blazoncek** [comment]
Closing as there has been no progress since December and code has progressed beyond simple resolving conflicts.

**Pull request closed without merging**